// @arwes/animated/build/cjs/createAnimatedXElement/createAnimatedXElement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAnimatedXElement = void 0;
const motion_1 = require("motion");
const tools_1 = require("@arwes/tools");
const index_js_1 = require("../applyAnimatedCSSProps/index.js");
const index_js_2 = require("../easing/index.js");
const createAnimatedXElement = (props) => {
    const { element } = props;
    const getSettings = () => ({
        hideOnStates: [],
        renderInitials: true,
        ...(0, tools_1.filterProps)(props.settingsRef.current)
    });
    let stateLastExecuted = props.settingsRef.current.state;
    const animations = new Set();
    const settingsInitial = getSettings();
    if (settingsInitial.renderInitials) {
        const { animated } = settingsInitial;
        const animatedListReceived = Array.isArray(animated) ? animated : [animated];
        const animatedList = animatedListReceived.filter(Boolean);
        const initialAttributes = animatedList
            .map((item) => item?.initialAttributes)
            .reduce((total, item) => ({ ...total, ...item }), {});
        Object.keys(initialAttributes).forEach((attribute) => {
            element.setAttribute(attribute, initialAttributes[attribute]);
        });
        const dynamicStyles = animatedList
            .map((item) => item?.initialStyle)
            .reduce((total, item) => ({ ...total, ...item }), {});
        (0, index_js_1.applyAnimatedCSSProps)(element, dynamicStyles);
    }
    const runAnimations = () => {
        const { state, animated, hideOnStates } = getSettings();
        stateLastExecuted = state;
        const animatedListReceived = Array.isArray(animated) ? animated : [animated];
        const animatedList = animatedListReceived.filter(Boolean);
        element.style.visibility = hideOnStates?.includes(state) ? 'hidden' : '';
        const $ = (query) => Array.from(element.querySelectorAll(query));
        animatedList
            .map((settingsItem) => settingsItem.transitions?.[state])
            .filter(Boolean)
            .forEach((transition) => {
            if (typeof transition === 'function') {
                const animation = transition({
                    element,
                    $,
                    easing: index_js_2.easing,
                    animate: motion_1.animate,
                    timeline: motion_1.timeline,
                    stagger: motion_1.stagger,
                    spring: motion_1.spring,
                    glide: motion_1.glide
                });
                if (animation) {
                    animations.add(animation);
                    if (animation.then) {
                        void animation.then(() => {
                            animations.delete(animation);
                        });
                    }
                    else if (animation.finished) {
                        void animation.finished.then(() => {
                            animations.delete(animation);
                        });
                    }
                }
            }
            else {
                const { duration, delay, easing: ease, repeat, direction, options, ...definition } = transition;
                if (Number.isFinite(duration) && duration <= 0) {
                    throw new Error('ARWES createAnimatedXElement() animation duration must be greater than 0.');
                }
                try {
                    const animation = (0, motion_1.animate)(element, definition, {
                        duration,
                        delay,
                        easing: typeof ease === 'string' ? index_js_2.easing[ease] : ease,
                        repeat,
                        direction,
                        ...options
                    });
                    animations.add(animation);
                    void animation.finished.then(() => {
                        animations.delete(animation);
                    });
                }
                catch (err) {
                    throw new Error(`ARWES createAnimatedXElement() animation error:\n${String(err)}`);
                }
            }
        });
    };
    runAnimations();
    const refresh = () => {
        if (stateLastExecuted !== props.settingsRef.current.state) {
            runAnimations();
        }
    };
    const cancel = () => {
        animations.forEach((animation) => animation.cancel());
        animations.clear();
    };
    return Object.freeze({ refresh, cancel });
};
exports.createAnimatedXElement = createAnimatedXElement;
