// @arwes/frames/build/cjs/createFrameNeroSettings/createFrameNeroSettings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createFrameNeroSettings = void 0;
const tools_1 = require("@arwes/tools");
const defaultProps = {
    styled: true,
    animated: true,
    padding: 0,
    cornerLength: 24,
    cornerWidth: 2
};
const createFrameNeroSettings = (props) => {
    const { styled, animated, padding: p, cornerWidth, cornerLength } = { ...defaultProps, ...(props ? (0, tools_1.filterProps)(props) : null) };
    const cornerHalf = cornerLength / 2;
    const cornerInternal = cornerLength / 2 - cornerWidth * 2;
    return {
        elements: [
            {
                type: 'rect',
                name: 'bg',
                x: p,
                y: p,
                width: `100% - ${p * 2}`,
                height: `100% - ${p * 2}`,
                style: {
                    filter: styled ? 'var(--arwes-frames-bg-filter)' : undefined,
                    fill: styled ? 'var(--arwes-frames-bg-color, currentcolor)' : undefined,
                    strokeWidth: styled ? '0' : undefined
                },
                animated: animated && {
                    initialStyle: { opacity: 0 },
                    transitions: {
                        entering: ({ element, duration, animate }) => animate(element, { opacity: 1 }, { delay: duration / 2, duration: duration / 2 }),
                        exiting: ({ element, duration, animate }) => animate(element, { opacity: 0 }, { duration: duration / 2 })
                    }
                }
            },
            {
                type: 'g',
                style: {
                    filter: styled ? 'var(--arwes-frames-deco-filter)' : undefined,
                    fill: styled ? 'var(--arwes-frames-deco-color, currentcolor)' : undefined,
                    strokeWidth: styled ? '0' : undefined
                },
                animated: animated && {
                    transitions: {
                        entering: ({ element, duration, animate }) => animate(element, { opacity: [0, 1, 0.5, 1] }, { duration: duration / 2 }),
                        exiting: ({ element, duration, animate }) => animate(element, { opacity: [1, 0, 0.5, 0] }, { delay: duration / 2, duration: duration / 2 })
                    }
                },
                elements: [
                    {
                        type: 'path',
                        name: 'deco',
                        animated: animated && [
                            ['x', cornerHalf, 0, undefined, 'outExpo'],
                            ['y', cornerHalf, 0, undefined, 'outExpo']
                        ],
                        path: [
                            ['M', p, p],
                            ['h', cornerLength],
                            ['l', -cornerWidth, cornerWidth],
                            ['h', -cornerHalf],
                            ['l', -cornerInternal, cornerInternal],
                            ['v', cornerHalf],
                            ['l', -cornerWidth, cornerWidth],
                            'z'
                        ]
                    },
                    {
                        type: 'path',
                        name: 'deco',
                        animated: animated && [
                            ['x', -cornerHalf, 0, undefined, 'outExpo'],
                            ['y', cornerHalf, 0, undefined, 'outExpo']
                        ],
                        path: [
                            ['M', `100% - ${p}`, p],
                            ['v', cornerLength],
                            ['l', -cornerWidth, -cornerWidth],
                            ['v', -cornerHalf],
                            ['l', -cornerInternal, -cornerInternal],
                            ['h', -cornerHalf],
                            ['l', -cornerWidth, -cornerWidth],
                            'z'
                        ]
                    },
                    {
                        type: 'path',
                        name: 'deco',
                        animated: animated && [
                            ['x', -cornerHalf, 0, undefined, 'outExpo'],
                            ['y', -cornerHalf, 0, undefined, 'outExpo']
                        ],
                        path: [
                            ['M', `100% - ${p}`, `100% - ${p}`],
                            ['h', -cornerLength],
                            ['l', cornerWidth, -cornerWidth],
                            ['h', cornerHalf],
                            ['l', cornerInternal, -cornerInternal],
                            ['v', -cornerHalf],
                            ['l', cornerWidth, -cornerWidth],
                            'z'
                        ]
                    },
                    {
                        type: 'path',
                        name: 'deco',
                        animated: animated && [
                            ['x', cornerHalf, 0, undefined, 'outExpo'],
                            ['y', -cornerHalf, 0, undefined, 'outExpo']
                        ],
                        path: [
                            ['M', p, `100% - ${p}`],
                            ['h', cornerLength],
                            ['l', -cornerWidth, -cornerWidth],
                            ['h', -cornerHalf],
                            ['l', -cornerInternal, -cornerInternal],
                            ['v', -cornerHalf],
                            ['l', -cornerWidth, -cornerWidth],
                            'z'
                        ]
                    }
                ]
            }
        ]
    };
};
exports.createFrameNeroSettings = createFrameNeroSettings;
