// @arwes/animator/build/cjs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ANIMATOR_DEFAULT_SETTINGS = exports.ANIMATOR_DEFAULT_DURATION = exports.ANIMATOR_MANAGER_NAMES = exports.ANIMATOR_ACTIONS = exports.ANIMATOR_STATES = void 0;
exports.ANIMATOR_STATES = Object.freeze({
    entered: 'entered',
    entering: 'entering',
    exiting: 'exiting',
    exited: 'exited'
});
exports.ANIMATOR_ACTIONS = Object.freeze({
    setup: 'setup',
    enter: 'enter',
    enterEnd: 'enterEnd',
    exit: 'exit',
    exitEnd: 'exitEnd',
    update: 'update',
    refresh: 'refresh'
});
exports.ANIMATOR_MANAGER_NAMES = Object.freeze({
    parallel: 'parallel',
    stagger: 'stagger',
    staggerReverse: 'staggerReverse',
    sequence: 'sequence',
    sequenceReverse: 'sequenceReverse',
    switch: 'switch'
});
exports.ANIMATOR_DEFAULT_DURATION = Object.freeze({
    enter: 0.4,
    exit: 0.4,
    delay: 0,
    offset: 0,
    stagger: 0.04,
    limit: -1
});
exports.ANIMATOR_DEFAULT_SETTINGS = Object.freeze({
    active: true,
    duration: exports.ANIMATOR_DEFAULT_DURATION,
    manager: exports.ANIMATOR_MANAGER_NAMES.parallel,
    merge: false,
    combine: false,
    initialState: 'exited'
});
