// @arwes/animated/build/cjs/transitions/transitions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.draw = exports.flicker = exports.fade = exports.transition = void 0;
const index_js_1 = require("../easing/index.js");
const index_js_2 = require("../animateDraw/index.js");
const transition = (prop, from, to, back, easing) => ({
    transitions: {
        entering: { [prop]: [from, to], easing },
        exiting: { [prop]: [to, back ?? from], easing }
    }
});
exports.transition = transition;
const fadeTransition = Object.freeze({
    transitions: {
        entering: { opacity: [0, 1] },
        exiting: { opacity: [1, 0] }
    }
});
const fade = () => fadeTransition;
exports.fade = fade;
const flickerTransition = Object.freeze({
    transitions: {
        entering: { opacity: [0, 1, 0.5, 1], easing: index_js_1.easing.outSine },
        exiting: { opacity: [1, 0, 0.5, 0], easing: index_js_1.easing.outSine }
    }
});
const flicker = () => flickerTransition;
exports.flicker = flicker;
const draw = (durationCustom, easing) => ({
    transitions: {
        entering: ({ element, duration }) => (0, index_js_2.animateDraw)({
            isEntering: true,
            element: element,
            duration: durationCustom ?? duration,
            easing
        }),
        exiting: ({ element, duration }) => (0, index_js_2.animateDraw)({
            isEntering: false,
            element: element,
            duration: durationCustom ?? duration,
            easing
        })
    }
});
exports.draw = draw;
