// @arwes/animated/build/esm/transitions/transitions.js
import { easing } from '../easing/index.js';
import { animateDraw } from '../animateDraw/index.js';
const transition = (prop, from, to, back, easing) => ({
    transitions: {
        entering: { [prop]: [from, to], easing },
        exiting: { [prop]: [to, back ?? from], easing }
    }
});
const fadeTransition = Object.freeze({
    transitions: {
        entering: { opacity: [0, 1] },
        exiting: { opacity: [1, 0] }
    }
});
const fade = () => fadeTransition;
const flickerTransition = Object.freeze({
    transitions: {
        entering: { opacity: [0, 1, 0.5, 1], easing: easing.outSine },
        exiting: { opacity: [1, 0, 0.5, 0], easing: easing.outSine }
    }
});
const flicker = () => flickerTransition;
const draw = (durationCustom, easing) => ({
    transitions: {
        entering: ({ element, duration }) => animateDraw({
            isEntering: true,
            element: element,
            duration: durationCustom ?? duration,
            easing
        }),
        exiting: ({ element, duration }) => animateDraw({
            isEntering: false,
            element: element,
            duration: durationCustom ?? duration,
            easing
        })
    }
});
export { transition, fade, flicker, draw };
