// @arwes/react-effects/build/esm/Illuminator/Illuminator.js
import React, { useRef, useEffect } from 'react';
import { cx } from '@arwes/tools';
import { memo, mergeRefs } from '@arwes/react-tools';
import { createEffectIlluminator } from '@arwes/effects';
const Illuminator = memo((props) => {
    const { elementRef, color, size, ...otherProps } = props;
    const containerRef = useRef(null);
    useEffect(() => {
        const container = containerRef.current;
        if (!container) {
            return;
        }
        const illuminator = createEffectIlluminator({ container, color, size });
        return () => illuminator.cancel();
    }, [color, size]);
    return (React.createElement("div", { ref: mergeRefs(containerRef, elementRef), role: "presentation", ...otherProps, className: cx('arwes-frames-illuminator', otherProps.className), style: {
            position: 'absolute',
            inset: 0,
            overflow: 'hidden',
            width: '100%',
            height: '100%',
            ...otherProps.style
        } }));
});
export { Illuminator };
