// @arwes/react-effects/build/cjs/Illuminator/Illuminator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Illuminator = void 0;
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importStar(require("react"));
const tools_1 = require("@arwes/tools");
const react_tools_1 = require("@arwes/react-tools");
const effects_1 = require("@arwes/effects");
const Illuminator = (0, react_tools_1.memo)((props) => {
    const { elementRef, color, size, ...otherProps } = props;
    const containerRef = (0, react_1.useRef)(null);
    (0, react_1.useEffect)(() => {
        const container = containerRef.current;
        if (!container) {
            return;
        }
        const illuminator = (0, effects_1.createEffectIlluminator)({ container, color, size });
        return () => illuminator.cancel();
    }, [color, size]);
    return (react_1.default.createElement("div", { ref: (0, react_tools_1.mergeRefs)(containerRef, elementRef), role: "presentation", ...otherProps, className: (0, tools_1.cx)('arwes-frames-illuminator', otherProps.className), style: {
            position: 'absolute',
            inset: 0,
            overflow: 'hidden',
            width: '100%',
            height: '100%',
            ...otherProps.style
        } }));
});
exports.Illuminator = Illuminator;
