// @arwes/animated/build/cjs/formatAnimatedCSSProps/formatAnimatedCSSProps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatAnimatedCSSProps = void 0;
const propsTransformDistances = {
    x: 'translateX',
    y: 'translateY',
    z: 'translateZ'
};
const propsTransformAngles = {
    rotate: 'rotate',
    rotateX: 'rotateX',
    rotateY: 'rotateY',
    rotateZ: 'rotateZ',
    skew: 'skew',
    skewX: 'skewX',
    skewY: 'skewY'
};
const propsTransformUnitless = {
    scale: 'scale',
    scaleX: 'scaleX',
    scaleY: 'scaleY',
    scaleZ: 'scaleZ'
};
const formatAnimatedCSSProps = (cssPropertiesEnhanced) => {
    const cssProperties = {};
    let transform = '';
    Object.keys(cssPropertiesEnhanced).forEach((key) => {
        const raw = cssPropertiesEnhanced[key];
        if (propsTransformDistances[key]) {
            const name = propsTransformDistances[key];
            const value = Number.isFinite(raw) ? `${raw}px` : String(raw);
            transform += ` ${name}(var(--motion-${name}))`;
            cssProperties[`--motion-${name}`] = value;
        }
        else if (propsTransformAngles[key]) {
            const name = propsTransformAngles[key];
            const value = Number.isFinite(raw) ? `${raw}deg` : String(raw);
            transform += ` ${name}(var(--motion-${name}))`;
            cssProperties[`--motion-${name}`] = value;
        }
        else if (propsTransformUnitless[key]) {
            const name = propsTransformUnitless[key];
            transform += ` ${name}(var(--motion-${name}))`;
            cssProperties[`--motion-${name}`] = raw;
        }
        else {
            ;
            cssProperties[key] = raw;
        }
    });
    transform = transform.trim();
    if (transform) {
        cssProperties.transform = transform;
    }
    return cssProperties;
};
exports.formatAnimatedCSSProps = formatAnimatedCSSProps;
